dvc/cache/base.py

```diff
@@ -623,77 +598,19 @@ def get_files_number(self, path_info, hash_info, filter_info):
             return 1
 
         if not filter_info:
-            return len(self.get_dir_cache(hash_info))
+            return self.get_dir_cache(hash_info).nfiles
 
         return ilen(
-            filter_info.isin_or_eq(path_info / entry[self.tree.PARAM_CHECKSUM])
-            for entry in self.get_dir_cache(hash_info)
-        )
-
-    def _to_dict(self, dir_info):
-        info = {}
-        for _entry in dir_info:
-            entry = _entry.copy()
-            relpath = entry.pop(self.tree.PARAM_RELPATH)
-            info[relpath] = HashInfo.from_dict(entry)
-        return info
-
-    def _from_dict(self, dir_dict):
-        return [
-            {self.tree.PARAM_RELPATH: relpath, **hash_info.to_dict()}
-            for relpath, hash_info in dir_dict.items()
-        ]
-
-    @staticmethod
-    def _diff(ancestor, other, allow_removed=False):
-        from dictdiffer import diff
-
-        allowed = ["add"]
-        if allow_removed:
-            allowed.append("remove")
-
-        result = list(diff(ancestor, other))
-        for typ, _, _ in result:
-            if typ not in allowed:
-                raise MergeError(
-                    "unable to auto-merge directories with diff that contains "
-                    f"'{typ}'ed files"
-                )
-        return result
-
-    def _merge_dirs(self, ancestor_info, our_info, their_info):
-        from dictdiffer import patch
-
-        ancestor = self._to_dict(ancestor_info)
-        our = self._to_dict(our_info)
-        their = self._to_dict(their_info)
-
-        our_diff = self._diff(ancestor, our)
-        if not our_diff:
-            return self._from_dict(their)
-
-        their_diff = self._diff(ancestor, their)
-        if not their_diff:
-            return self._from_dict(our)
-
-        # make sure there are no conflicting files
-        self._diff(our, their, allow_removed=True)
-
-        merged = patch(our_diff + their_diff, ancestor, in_place=True)
-
-        # Sorting the list by path to ensure reproducibility
-        return sorted(
-            self._from_dict(merged), key=itemgetter(self.tree.PARAM_RELPATH),
+            filter_info.isin_or_eq(path_info / relpath)
+            for relpath, _ in self.get_dir_cache(hash_info).items()
         )
 
     def _get_dir_size(self, dir_info):
-        def _getsize(entry):
-            return self.tree.getsize(
-                self.tree.hash_to_path_info(entry[self.tree.PARAM_CHECKSUM])
-            )
-
         try:
-            return sum(_getsize(entry) for entry in dir_info)
+            return sum(
+                self.tree.getsize(self.tree.hash_to_path_info(hi.value))
+                for _, hi in dir_info.items()
+            )
         except FileNotFoundError:
             return None
```

Comment:
Ideally `size` should be added to dir_info entries when loading it, so that we could simply use DirInfo.size instead of doing this manual labor. Will look into it...

Reply:
Ok, not at this point. Keeping as is for now.